src/returns.rs

```diff
@@ -0,0 +1,100 @@
+
+#[derive(Copy,Clone)]
+pub struct ReturnPass;
+
+impl ReturnPass {
+    // Check the final stmt or expr in a block for unnecessary return.
+    fn check_block_return(&mut self, cx: &Context, block: &Block) {
+        if let Some(ref expr) = block.expr {
+            self.check_final_expr(cx, expr);
+        } else if let Some(stmt) = block.stmts.last() {
+            if let StmtSemi(ref expr, _) = stmt.node {
+                if let ExprRet(Some(ref inner)) = expr.node {
+                    self.emit_return_lint(cx, (expr.span, inner.span));
+                }
+            }
+        }
+    }
+
+    // Check a the final expression in a block if it's a return.
+    fn check_final_expr(&mut self, cx: &Context, expr: &Expr) {
+        match expr.node {
+            // simple return is always "bad"
+            ExprRet(Some(ref inner)) => {
+                self.emit_return_lint(cx, (expr.span, inner.span));
+            }
+            // a whole block? check it!
+            ExprBlock(ref block) => {
+                self.check_block_return(cx, block);
+            }
+            // an if/if let expr, check both exprs
+            // note, if without else is going to be a type checking error anyways
+            // (except for unit type functions) so we don't match it
+            ExprIf(_, ref ifblock, Some(ref elsexpr)) |
+            ExprIfLet(_, _, ref ifblock, Some(ref elsexpr)) => {
+                self.check_block_return(cx, ifblock);
+                self.check_final_expr(cx, elsexpr);
+            }
+            // a match expr, check all arms
+            ExprMatch(_, ref arms, _) => {
+                for arm in arms {
+                    self.check_final_expr(cx, &*arm.body);
+                }
+            }
+            _ => { }
+        }
+    }
+
+    fn emit_return_lint(&mut self, cx: &Context, spans: (Span, Span)) {
+        span_lint(cx, NEEDLESS_RETURN, spans.0, &format!(
+            "unneeded return statement. Consider using {} \
+             without the trailing semicolon",
+            snippet(cx, spans.1, "..")))
+    }
+
+    // Check for "let x = EXPR; x"
+    fn check_let_return(&mut self, cx: &Context, block: &Block) {
+        // we need both a let-binding stmt and an expr
+        if let Some(stmt) = block.stmts.last() {
+            if let StmtDecl(ref decl, _) = stmt.node {
+                if let DeclLocal(ref local) = decl.node {
+                    if let PatIdent(_, Spanned { node: id, .. }, _) = local.pat.node {
+                        if let Some(ref retexpr) = block.expr {
+                            if let ExprPath(_, ref path) = retexpr.node {
+                                if match_path(path, &[&*id.name.as_str()]) {
+                                    span_lint(cx, LET_AND_RETURN, retexpr.span,
+                                              "returning the result of a let binding. \
+                                               Consider returning the expression directly.");
```

Comment:
Add a `note: this expression can be directly returned` below which points to the expr.span